### DISCUSSION TITLE
How to extend prisma model types to include properties added by a client extension?

### DISCUSSION THREAD

**Discussion by ludwigbacklund** [prisma/prisma#18460] [Q&A] [answered]
I add a `__typename` property to all my prisma model objects via a client extension ([it looks like this](https://github.com/prisma/prisma/issues/5315#issuecomment-1447116621), but with `__typename` instead of `$kind`). However, when I import say the `User` model type, it does not have this `__typename` property in the type.

How do I extend the `User` model type to know that it now has a `__typename` property on it? I suspect that these model types that I'm importing are describing the database model structure of a `User`, and not what I get from my extended prisma client. Is there perhaps a different way I can reference the `User` model type, but as it would look coming out of my extended prisma client?

One way I could see this working is being able to do something like `ExtendedPrismaClient['User']`, but nothing like that exists on the extended prisma client type today.

Thank you!

**Comment by ludralph** [comment]
Hi @ludwigbacklund 👋 

Can you please share the code for your extension and how you are importing it?

**Reply by ludwigbacklund** [reply]
Extension:

```ts
const brandExtension = Prisma.defineExtension((client) => {
  type ModelKey = Exclude<keyof typeof client, `$${string}` | symbol>;
  type Result = {
    [K in ModelKey]: { __typename: { needs: Record<string, never>; compute: () => K } };
  };

  const result = {} as Result;
  const modelKeys = Object.keys(client).filter((key) => !key.startsWith('$')) as ModelKey[];
  modelKeys.forEach((k) => {
    result[k] = { __typename: { needs: {}, compute: () => k as any } };
  });

  return client.$extends({ result });
});

const client = new PrismaClient({
  datasources: { db: { url: 'DATABASE_URL' } },
  log: ['error'],
}).$extends(brandExtension);
```

Import:

```ts
//        👇 User does not have `__typename`
import { User } from '@prisma/client';
```

**Reply by millsp** [reply, marked as answer]
Hey :wave: It is not possible to change that. The imported types are "immutable" and static, meaning that an extension will not change their defined shape. We don't currently provide an easy or stable way to pass extension info into the types, but we could in the future. Would you mind opening a feature request?

Right now I suggest that you create a custom type manually:

```ts
type UserBrand = User & { __typename: "user" }
```

**Reply by nurul3101** [reply]
@ludwigbacklund If this answers your question, it would be great if you could **mark this Discussion as answered** to indicate that it has been resolved. 

Otherwise please let us know how else we can help you further or close the Discussion if it was resolved in some other way 🙏

**Reply by ludwigbacklund** [reply]
Thank you. I submitted a feature request here: https://github.com/prisma/prisma/issues/19079

**Reply by nurul3101** [reply]
Thanks for sharing the feature request 👍

**Comment by ludwigbacklund** [comment]
If anyone has insights here, it would be appreciated!
